requests/transport.go

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (C) 2019 The "MysteriumNetwork/node" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package requests
+
+import (
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/rs/zerolog/log"
+)
+
+const (
+	maxHTTPRetries  = 3
+	delayAfterRetry = 50 * time.Millisecond
+)
+
+func newTransport(localIPAddress *net.TCPAddr) *transport {
+	t := &transport{
+		rt: &http.Transport{
+			DialContext: (&net.Dialer{
+				LocalAddr: localIPAddress,
+			}).DialContext,
+		},
+	}
+	return t
+}
+
+type transport struct {
+	rt http.RoundTripper
+}
+
+// RoundTrip does HTTP requests with retries. Retries are needed since
+// error can happen when HTTP client closes idle connections.
+func (t *transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	for i := 1; i <= maxHTTPRetries; i++ {
+		res, err := t.rt.RoundTrip(req)
+		if err == nil {
+			return res, nil
+		}
+
+		log.Error().Err(err).Msgf("Failed to call %q. Retrying.", req.URL.String())
+		time.Sleep(delayAfterRetry)
```

Comment:
this prevents graceful shutdowns. Say I stop the node, the connection state changes to cancelled. This cancels the request which then causes this request to be retries 2 more times. If the requests are slow, it will hang for a long imte. It will cause the node to hang. Suggest subscribing to a stop event, which will then cause a channel to be closed. This part can then become something like this:

```golang
select {
    case <- time.After(delayAfterRetry):
                 if i == maxHTTPRetries {
			return res, err
		}
   case <-t.stop:
               return nil, nil
} 
```

Reply:
I tested in with this simple program.

```go
func runHTTP() {
	c := http.Client{
		Timeout:   20 * time.Second,
		Transport: newTransport(),
	}

	go func() {
		panic(http.ListenAndServe(":8005", nil))
	}()

	for {
		req, _ := http.NewRequest("GET", "http://localhost:8007", nil)
		res, err := c.Do(req)
		if err != nil {
			fmt.Printf("ERR: %v\n", err)
			continue
		}
		defer res.Body.Close()
		b, _ := ioutil.ReadAll(res.Body)
		fmt.Println("OK", time.Now().String(), len(b))
	}
}

func newTransport() {
   // Uses the same logic as is now.
}
```

- Client calls some HTTP server on localhost:8007. Server exposes slow HTTP endpoint which returns response only after 10 seconds.
- Client also starts HTTP on localhost:8005. After I stop client I can run it again immediately otherwise I should get error saying that HTTP port 8005 is in use.

Reply:
Added changes as suggested.